File: dataflow-sdk/src/main/java/tk/fishfish/dataflow/sdk/domain/Graph.java
package tk.fishfish.dataflow.sdk.domain;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.List;
import java.util.Map;

/**
 * 图
 *
 * @author 奔波儿灞
 * @version 1.0.0
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Graph {

    private List<Node> nodes;

    private List<Edge> edges;

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Node {

        private String id;

        private String name;

        private String text;

        private Argument argument;

    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Edge {

        private String from;

        private String to;

    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class Argument {

        private Map<String, Object> input;

        private Map<String, Object> output;

    }

}

File: dataflow-core/src/main/scala/tk/fishfish/dataflow/dag/Graph.scala
package tk.fishfish.dataflow.dag

import org.apache.commons.lang3.StringUtils
import tk.fishfish.dataflow.task.Argument

import java.util.Objects

/**
 * 任务流
 *
 * @author 奔波儿灞
 * @version 1.0.0
 */
case class Graph(nodes: Seq[Node], edges: Seq[Edge])

case class Node(id: String, name: String, text: String, argument: Argument) {

  override def hashCode(): Int = Objects.hash(id)

  override def equals(obj: Any): Boolean = {
    if (obj == null || !obj.isInstanceOf[Node]) {
      return false
    }
    StringUtils.equals(this.id, obj.asInstanceOf[Node].id)
  }

}

case class Edge(from: String, to: String) {

  override def hashCode(): Int = Objects.hash(from, to)

  override def equals(obj: Any): Boolean = {
    if (obj == null || !obj.isInstanceOf[Edge]) {
      return false
    }
    val edge = obj.asInstanceOf[Edge]
    StringUtils.equals(this.from, edge.from) && StringUtils.equals(this.to, edge.to)
  }

}
